File: RightViewDFS.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
class Solution {
    List<Integer> result;
    public List<Integer> rightSideView(TreeNode root) {
        result = new LinkedList<>();
        
        if(root == null){
            return result;
        }
        
        dfs(root, 0);
        return result;
    }
    
    private void dfs(TreeNode root, int level){
        //base
        if(root == null){
            return;
        }
        
        //logic
        //Add the element on the left and replace it with next element on the same level
        if(level == result.size()){
            result.add(root.val);
        }else{
            result.set(level, root.val);
        }
        
        //Go to the left first 
        dfs(root.left, level+1);
        
        //Go to the right to replace the left elements
        dfs(root.right, level+1);
    }
}

File: RightViewDFS.py
#Time Complexity:O(n)
#Space Complexity:O(h)
#Check if the element exists at particular height, if exists skip else add to the output list
#Traversing root and then left

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    global output
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.output=[]
        self.DFS(root,0)
        return self.output

    def DFS(self, root, depth):
        if root is None:
            return
        if depth>=len(self.output):
            self.output.append(root.val)
        self.DFS(root.right,depth+1)
        self.DFS(root.left,depth+1)